akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala

```diff
@@ -693,3 +699,423 @@ private[akka] class BroadcastHub[T](bufferSize: Int) extends GraphStageWithMater
+    private def isFull: Boolean = {
+      (queue.size + pending.size) >= bufferSize
+    }
+
+    private def publish(elem: T): Unit = {
+      if (!initialized || consumers.isEmpty) {
+        // will be published when first consumers are registered
+        pending :+= elem
+      } else {
+        val p = partitioner(elem)
+        val id = consumers(p % consumers.size).id
+        queue.offer(id, elem)
+        wakeup(id)
+      }
+    }
+
+    private def wakeup(id: Long): Unit = {
+      needWakeup.get(id) match {
+        case None ⇒
+        case Some(consumer) ⇒
+          needWakeup -= id
+          consumer.callback.invoke(Wakeup)
+      }
+    }
+
+    override def onUpstreamFinish(): Unit = {
+      if (consumers.isEmpty)
+        completeStage()
+      else {
+        consumers.foreach(c ⇒ complete(c.id))
+      }
+    }
+
+    private def complete(id: Long): Unit = {
+      queue.offer(id, Completed)
+      wakeup(id)
+    }
+
+    private def tryPull(): Unit = {
+      if (initialized && !isClosed(in) && !hasBeenPulled(in) && !isFull)
+        pull(in)
+    }
+
+    private def onEvent(ev: HubEvent): Unit = {
+      callbackCount += 1
+      ev match {
+        case NeedWakeup(consumer) ⇒
+          // Also check if the consumer is now unblocked since we published an element since it went asleep.
+          if (queue.nonEmpty(consumer.id))
+            consumer.callback.invoke(Wakeup)
+          else {
+            needWakeup.update(consumer.id, consumer)
+            tryPull()
+          }
+
+        case TryPull ⇒
+          tryPull()
+
+        case RegistrationPending ⇒
+          state.getAndSet(noRegistrationsState).asInstanceOf[Open].registrations foreach { consumer ⇒
+            consumers :+= consumer
+            consumers = consumers.sortBy(_.id)
+            queue.init(consumer.id)
+            if (consumers.size >= startAfterNbrOfStreams) {
+              initialized = true
+            }
+
+            consumer.callback.invoke(Initialize)
+
+            if (initialized && pending.nonEmpty) {
+              pending.foreach(publish)
+              pending = Vector.empty[T]
+            }
+
+            tryPull()
+          }
+
+        case UnRegister(id) ⇒
+          consumers = consumers.filterNot(_.id == id)
+          queue.remove(id)
+          // FIXME completed?
+          if (consumers.isEmpty) {
+            if (isClosed(in)) completeStage()
+          } else
+            tryPull()
+      }
+    }
+
+    override def onUpstreamFailure(ex: Throwable): Unit = {
+      val failMessage = HubCompleted(Some(ex))
+
+      // Notify pending consumers and set tombstone
+      state.getAndSet(Closed(Some(ex))).asInstanceOf[Open].registrations foreach { consumer ⇒
+        consumer.callback.invoke(failMessage)
+      }
+
+      // Notify registered consumers
+      consumers.foreach { consumer ⇒
+        consumer.callback.invoke(failMessage)
+      }
+      failStage(ex)
+    }
+
+    override def postStop(): Unit = {
```

Comment:
sanity check, termination of hub happens when upstream completes, then we notify all downstreams using the callbacks